src/rust/engine/src/externs/target.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::fmt::Write;
+
+use pyo3::basic::CompareOp;
+use pyo3::exceptions::PyValueError;
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::PyType;
+
+use crate::externs::address::Address;
+
+pub fn register(m: &PyModule) -> PyResult<()> {
+  m.add_class::<Field>()?;
+  m.add_class::<NoFieldValue>()?;
+
+  m.add("NO_VALUE", NoFieldValue)?;
+
+  Ok(())
+}
+
+#[pyclass(name = "_NoValue")]
+#[derive(Clone)]
+struct NoFieldValue;
+
+#[pymethods]
+impl NoFieldValue {
+  fn __bool__(&self) -> bool {
+    false
+  }
+
+  fn __repr__(&self) -> &'static str {
+    "<NO_VALUE>"
+  }
+}
+
+#[pyclass(subclass)]
+pub struct Field {
+  value: PyObject,
+}
+
+#[pymethods]
+impl Field {
+  #[new]
+  #[classmethod]
+  #[pyo3(signature = (raw_value, address))]
+  fn __new__(
+    cls: &PyType,
+    raw_value: Option<PyObject>,
+    address: PyRef<Address>,
+    py: Python,
+  ) -> PyResult<Self> {
+    let raw_value = match raw_value {
+      Some(value)
+        if value.extract::<NoFieldValue>(py).is_ok() && !Self::cls_none_is_valid_value(cls)? =>
+      {
+        None
+      }
+      rv => rv,
+    };
+
+    Self::check_deprecated(cls, raw_value.as_ref(), &address, py)?;
+
+    Ok(Self {
+      value: cls
+        .call_method(intern!(py, "compute_value"), (raw_value, address), None)?
+        .into(),
+    })
+  }
+
+  #[classattr]
+  fn none_is_valid_value() -> bool {
+    false
+  }
+
+  #[classattr]
+  fn required() -> bool {
+    false
+  }
+
+  #[classattr]
+  fn removal_version() -> Option<&'static str> {
+    None
+  }
+
+  #[classattr]
+  fn removal_hint() -> Option<&'static str> {
+    None
+  }
+
+  #[classattr]
+  fn deprecated_alias() -> Option<&'static str> {
+    None
+  }
+
+  #[classattr]
+  fn deprecated_alias_removal_version() -> Option<&'static str> {
+    None
+  }
+
+  #[classmethod]
+  #[pyo3(signature = (raw_value, address))]
+  fn compute_value(
+    cls: &PyType,
+    raw_value: Option<PyObject>,
+    address: PyRef<Address>,
+    py: Python,
+  ) -> PyResult<PyObject> {
+    let default = || -> PyResult<PyObject> {
+      if Self::cls_required(cls)? {
+        // TODO: Should be `RequiredFieldMissingException`.
+        Err(PyValueError::new_err(format!(
```

Comment:
Is using an error defined in python more difficult than using the built in errors? This is the only conerning thing I see in this PR (from my post-review merge). Very cool stuff!